File: Algorithms_Part_2/Ex7DepthFirstSearchDigraph.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.Stack;

public class Ex7DepthFirstSearchDigraph {

    private boolean[] marked;
    private int[] edgeTo;

    public Ex7DepthFirstSearchDigraph(Ex6Digraph G, int s) {
        marked = new boolean[G.V()];
        edgeTo = new int[G.V()];
        for (int i = 0; i < G.V(); i++) {
            marked[i] = false;
            edgeTo[i] = -1;
        }

        dfs(G, s);
    }

    public boolean hasPathTo(int v) {
        return marked[v];
    }

    public Iterable<Integer> pathTo(int v) {
        Stack<Integer> path = new Stack<>();
        path.push(v);
        for (int w = v; edgeTo[w] != -1; w = edgeTo[w])
            path.push(edgeTo[w]);
        return path;
    }

    private void dfs(Ex6Digraph G, int v) {
        marked[v] = true;
        for (int w : G.adj(v)) {
            if (!marked[w]) {
                dfs(G, w);
                edgeTo[w] = v;
            }
        }
    }


    public static void main(String[] args) {

    }
}

File: PythonLearnig/Algorithms_Part_2/Ex7DepthFirstSearchDigraph.py
from Ex6Digraph import Digraph


class DepthFirstSearchDigraph:

    def __init__(self, digraph, source):

        def dfs(vertex):
            self.marked[vertex] = True
            for adj in digraph.get_adj(vertex):
                if not self.marked[adj]:
                    self.edge_to[adj] = vertex
                    dfs(adj)

        self.marked = [False] * digraph.get_v()
        self.edge_to = [-1] * digraph.get_v()
        if isinstance(source, list):
            for i in range(len(source)):
                if not self.marked[i]:
                    dfs(source[i])
        else:
            dfs(source)

    def has_path_to(self, v):
        return self.marked[v]

    def path_to(self, v):
        if self.edge_to[v] != -1:
            for path in self.path_to(self.edge_to[v]):
                yield path
        yield v


def test_depth_first_search_digraph():
    digraph = Digraph(0)
    digraph.init_from('digraph.txt')

    print()
    print(digraph)

    print()
    dfs = DepthFirstSearchDigraph(digraph, 0)
    for v in range(digraph.get_v()):
        message = f'Has Path to {v}? {dfs.has_path_to(v)}'
        if dfs.has_path_to(v):
            message += ' Path:'
            for p in dfs.path_to(v):
                message += f' {p}'
        print(message)


if __name__ == '__main__':
    test_depth_first_search_digraph()
